### PR TITLE
Feature/kylem/410 old requests

### PR DESCRIPTION
**Pull request by kmorrison**
- Add 410 functionality to endpoint and swagger
- Add two tests for single scope and zero scope expecting 410
- Remove dependencies that `changes_api_test` tests has between one another
- Rename `changes_api_test` `It`s to be more BDD

### CODE REVIEW

File: changeserver/changes_api_test.go [hunk 1]
```diff
@@ -106,9 +116,56 @@ var _ = Describe("Changes API Tests", func() {
...
+		// Delete everything older than one second from now, ie. everything
+		testServer.cleanOnce(-1 * time.Second)
+		resp, err := http.Get(url)
+		defer resp.Body.Close()
+		Expect(err).Should(Succeed())
+		Expect(resp.StatusCode).Should(Equal(410))
 	})
 
-	It("Snapshot filter", func() {
+	It("Return 410 when an input seq has been purged, even if scope is wider", func() {
```
Comment (kriswehner) [inline changeserver/changes_api_test.go]: Can we start adding "Should" to the front of these test names so they read correctly? Like It("should return 410 when an..."

File: changeserver/changes_api_test.go [hunk 2]
```diff
@@ -106,9 +116,56 @@ var _ = Describe("Changes API Tests", func() {
...
+			return false
+		}).Should(BeTrue())
+
+		url = fmt.Sprintf(
+			"%s/changes?since=%s", baseURL, startingSequence)
+
+		// Delete everything older than one second from now, ie. everything
+		testServer.cleanOnce(-1 * time.Second)
+		resp, err := http.Get(url)
+		defer resp.Body.Close()
```
Comment (kriswehner) [inline changeserver/changes_api_test.go]: You should do this after the error check. Doing it before is a fast path to having an invalid Body and crashing if there *is* an error.

File: changeserver/changes_api_test.go [hunk 3]
```diff
@@ -315,15 +372,17 @@ var _ = Describe("Changes API Tests", func() {
...
-	It("Long polling two scopes", func() {
+	It("Should poll then return a change when a change gets made while its polling for two scopes", func() {
+		lastTestSequence++
+		startingSequence := makeChange("foo", lastTestSequence)
 		resultChan := make(chan *common.ChangeList, 1)
 
 		go func() {
 			lc := getChanges(fmt.Sprintf(
-				"%s/changes?since=%s&scope=foo&scope=bar", baseURL, lastChangeSequence))
+				"%s/changes?since=%s&scope=foo&scope=bar", baseURL, startingSequence))
```
Comment (kriswehner) [inline changeserver/changes_api_test.go]: Neat. I just learned that sequence has a String() method.

File: changeserver/changesapi.go [hunk 4]
```diff
@@ -71,6 +71,30 @@ func (s *server) handleGetChanges(resp http.ResponseWriter, req *http.Request) {
 		snapshotFilter = makeSnapshotFilter(snapshot)
 	}
 
+	// Before we start trying to read changes, let's make sure the database is in a sane state by checking that the
+	// change that is requested is still in the database. If it's been deleted, it means that other changes we're
+	// supposed to sync might also have been deleted and that any further syncing from this point could skip changes.
+	changeNotFound := true
+	defaultSequence := common.Sequence{}
```
Comment (kriswehner) [inline changeserver/changesapi.go]: Nit: I think this is typically called "empty" in go

File: changeserver/changesapi.go [hunk 5]
```diff
@@ -71,6 +71,30 @@ func (s *server) handleGetChanges(resp http.ResponseWriter, req *http.Request) {
...
+	if sinceSeq != defaultSequence {
+		for _, scope := range scopes {
+			changeInDB, err := s.db.GetEntry(scope, sinceSeq.LSN, sinceSeq.Index)
+			if err != nil {
+				//log.Infof("TODO: Throw some fatal error")
```
Comment (kriswehner) [inline changeserver/changesapi.go]: Missing code?

Comment (kriswehner) [inline changeserver/changesapi.go]: FYI: If this is directly in the web service loop, you can just panic(err), that will cause the server to return a 500.

File: changeserver/changes_api_test.go [hunk 6]
```diff
@@ -106,9 +116,56 @@ var _ = Describe("Changes API Tests", func() {
...
 			return false
 		}).Should(BeTrue())
+
+		url = fmt.Sprintf(
+			"%s/changes?since=%s&scope=foo", baseURL, startingSequence)
+
+		// Delete everything older than one second from now, ie. everything
+		testServer.cleanOnce(-1 * time.Second)
+		resp, err := http.Get(url)
+		defer resp.Body.Close()
```
Comment (kriswehner) [inline changeserver/changes_api_test.go]: Same. Should be after the error check.

File: changeserver/changesapi.go [hunk 7]
```diff
@@ -75,12 +75,12 @@ func (s *server) handleGetChanges(resp http.ResponseWriter, req *http.Request) {
...
-	if sinceSeq != defaultSequence {
+	emptySequence := common.Sequence{}
+	if sinceSeq != emptySequence {
 		for _, scope := range scopes {
 			changeInDB, err := s.db.GetEntry(scope, sinceSeq.LSN, sinceSeq.Index)
 			if err != nil {
-				//log.Infof("TODO: Throw some fatal error")
+				sendError(resp, req, http.StatusInternalServerError, err.Error())
 			} else if len(changeInDB) == 0 {
 				// Checking other scopes, might be fine
```
Comment (kriswehner) [inline changeserver/changesapi.go]: Should this branch *do* something? If not, I vote remove it.

**Comment by kmorrison** [comment]
I will change the comment to be more clear. That comment is trying to explain that we can't take any action in that if-block yet, but is doing a poor job of it.
